src/keyring/KeyringController.ts

```diff
@@ -421,6 +433,25 @@ export class KeyringController extends BaseController<
   ) {
     try {
       const address = normalizeAddress(messageParams.from);
+      const qrKeyring = await this.getOrAddQRKeyring();
+      const qrAccounts = await qrKeyring.getAccounts();
+      if (
+        qrAccounts.findIndex(
+          (qrAddress: string) =>
+            qrAddress.toLowerCase() === address.toLowerCase(),
+        ) !== -1
+      ) {
+        const messageParamsClone = { ...messageParams };
+        if (
+          version !== SignTypedDataVersion.V1 &&
+          typeof messageParamsClone.data === 'string'
+        ) {
+          messageParamsClone.data = JSON.parse(messageParamsClone.data);
+        }
+        return privates
+          .get(this)
+          .keyring.signTypedMessage(messageParamsClone, { version });
+      }
```

Comment:
The method `signTypedMessage` is getting to big, in a next PR we should consider breaking it down to smaller internal methods

Reply:
With the experience I have in MM extension integration before, I think the typed message signing can be hold by keyring implementations. Maybe it is better to create another controller called HardwareController later to handle things with hardwares that we cannot get the private key directly.